File: packages/cockpit-protocol/src/remote-wire-magic-registry.ts
export interface RemoteWireMagicOwnerV1 {
  magic: string;
  symbolicType: string;
  owningPackage: string;
  owningVersion: number;
}
export function parseRemoteWireMagicRegistry(value: unknown): readonly RemoteWireMagicOwnerV1[] {
  if (!Array.isArray(value) || value.length === 0)
    throw new Error("magic registry must be nonempty");
  const magics = new Set<string>(),
    types = new Set<string>();
  let previous = "";
  return value.map((raw) => {
    if (!raw || typeof raw !== "object" || Array.isArray(raw))
      throw new Error("invalid registry entry");
    const x = raw as Record<string, unknown>;
    if (
      Object.keys(x).sort().join(",") !== "magic,owningPackage,owningVersion,symbolicType" ||
      typeof x.magic !== "string" ||
      !/^FC[A-Z0-9]{2}$/.test(x.magic) ||
      typeof x.symbolicType !== "string" ||
      !x.symbolicType ||
      typeof x.owningPackage !== "string" ||
      !x.owningPackage ||
      x.owningVersion !== 1
    )
      throw new Error("invalid registry entry");
    if (x.magic <= previous) throw new Error("registry not sorted or duplicate magic");
    if (types.has(x.symbolicType)) throw new Error("duplicate symbolic type");
    previous = x.magic;
    magics.add(x.magic);
    types.add(x.symbolicType);
    return x as unknown as RemoteWireMagicOwnerV1;
  });
}
export function assertRegisteredProductionMagics(
  registry: readonly RemoteWireMagicOwnerV1[],
  declared: readonly { magic: string; symbolicType: string }[],
) {
  const owners = new Map(registry.map((x) => [x.magic, x.symbolicType]));
  for (const x of declared)
    if (owners.get(x.magic) !== x.symbolicType)
      throw new Error(`unregistered production codec ${x.magic}`);
}

File: crates/cockpit-proto/src/remote_wire_magic_registry.rs
//! Strict parser for the shared remote binary-magic namespace.
use serde::Deserialize;
use std::collections::HashSet;

#[derive(Debug, Clone, PartialEq, Eq, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct MagicOwner {
    pub magic: String,
    pub symbolic_type: String,
    pub owning_package: String,
    pub owning_version: u8,
}
pub fn parse_registry(json: &str) -> Result<Vec<MagicOwner>, String> {
    let xs: Vec<MagicOwner> = serde_json::from_str(json).map_err(|e| e.to_string())?;
    if xs.is_empty() {
        return Err("magic registry must be nonempty".into());
    }
    let mut m = HashSet::new();
    let mut t = HashSet::new();
    let mut previous = "";
    for x in &xs {
        if x.magic.len() != 4
            || !x.magic.starts_with("FC")
            || !x.magic[2..]
                .bytes()
                .all(|b| b.is_ascii_uppercase() || b.is_ascii_digit())
            || x.magic.as_str() <= previous
            || !m.insert(x.magic.clone())
            || x.symbolic_type.is_empty()
            || !t.insert(x.symbolic_type.clone())
            || x.owning_package.is_empty()
            || x.owning_version != 1
        {
            return Err("invalid, unsorted, duplicate, or ownerless magic entry".into());
        }
        previous = &x.magic;
    }
    Ok(xs)
}
pub fn assert_registered(registry: &[MagicOwner], declared: &[(&str, &str)]) -> Result<(), String> {
    for (magic, symbol) in declared {
        if !registry
            .iter()
            .any(|x| x.magic == *magic && x.symbolic_type == *symbol)
        {
            return Err(format!("unregistered production codec {magic}"));
        }
    }
    Ok(())
}
